Fix block.py and add test for file block.py

hashbangstudio opened the pull request:
For Block class:
  Add docstrings for functions in block.py
  Fix references to attribute so all self.type not split between self.type and self.id
  Add id property to maintain backwards compatibility
  type and data are always integers and should be treated as such in `__repr__`
  As `__cmp__` no longer in Python 3 need to use functools.total_ordering which is in 2.7 and 3.*
  https://docs.python.org/2/library/functools.html#functools.total_ordering
  https://docs.python.org/3/library/functools.html#functools.total_ordering
  Define `__eq__` and `__lt__` and others are created automatically

For blockType enumeration:
  Move to back ported IntEnum as allows direct integer comparison
  https://pypi.python.org/pypi/enum34
  https://docs.python.org/3/library/enum.html#intenum

Add test_block.py:
  A unit test module to test the block file

hashbangstudio committed:
Fix block.py and add test for file

For Block class:
  Add docstrings for functions in block.py
  Fix references to attribute so all self.type not split between self.type and self.id
  Add id property to maintain backwards compatibility
  type and data are always integers and should be treated as such in __repr__
  As __cmp__ no longer in Python 3 need to use functools.total_ordering which is in 2.7 and 3.*
  https://docs.python.org/2/library/functools.html#functools.total_ordering
  https://docs.python.org/3/library/functools.html#functools.total_ordering
  Define __eq__ and __lt__ and others are created automatically

For blockType enumeration:
  Move to back ported IntEnum as allows direct integer comparison
  https://pypi.python.org/pypi/enum34
  https://docs.python.org/3/library/enum.html#intenum

Add test_block.py:
  A unit test module to test the block file

doismellburning commented:
Why are blocks orderable? Does some odd bit of the API depend on this? :s

hashbangstudio commented:
Don't have to orderable as far as I am aware but they had been defined as such with `__cmp__` in the original API and thought safest to replicate the original behaviour. Could just define `__eq__` and `__ne__` if people prefer.

hashbangstudio commented:
Until enum34 package defined in Travis,  build will fail see #23 as enum won't exist.

doismellburning commented:
I don't think nonsensical orderings are helpful - will look to see if it's actually used anywhere (I hope not...)

doismellburning commented on tests/test_block.py:
```diff
@@ -0,0 +1,130 @@
+        for index, attr in enumerate(b):
+            self.assertEqual(attr, idAndData[index])
+
+    def testBlockConstants(self):
```
I may be alone here, but I'm unconvinced by the value of testing constants - persuade me...?

doismellburning commented on tests/test_block.py:
```diff
@@ -0,0 +1,130 @@
+
+class TestBlock(unittest.TestCase):
+
+    def testRepresentation(self):
```
Is this not just testing that "the string format we hardcoded over there is the same we hardcoded over here"?

I'm not sure about the existence/state of QuickCheck for Python, but perhaps lets (attempt to) instead validate the `repr`/`eval` identity?

hashbangstudio commented:
Testing that the name exists as well as the value in this case with enum.
Thinking was in case a line in the blockType definition gets deleted/altered, or name has been changed accidentally from e.g. global find and replace.
In both cases enum is still valid to interpreter, but one element of enum that would be used in code won't exist.
Should always be caught on commit but have seen mistakes happen or all modifications committed unintentionally.
Additionally are checking the aliasing of water and lava to flowing variants.
Easy enough to lose if unwanted.

doismellburning committed:
Added missing enum34 dependency

hashbangstudio committed:
Rename test to indicate not py.test form

doismellburning committed:
Revert "Added missing enum34 dependency"

This reverts commit ff8a4643eea128bfb5ba04f12c4425320771e364.

Oops, it's a python package not a deb!

hashbangstudio committed:
Remove runner code in test

hashbangstudio committed:
Remove block ordering comparisons

hashbangstudio committed:
Remove ordered comparison tests

dbrgn commented on minecraft/block.py:
```diff
@@ -1,42 +1,68 @@
+        block.data = The variant of the type of block.
+        For example the colour of wool or the orientation of stairs
+        The default type for blocks is dirt
+        """
```
Why the indentation change?

ghickman replied:
Yes, this is unnecessary, can you keep the indentation of docstrings to one level in line with [PEP 257](http://legacy.python.org/dev/peps/pep-0257/#handling-docstring-indentation) please.

doismellburning replied:
@ghickman Got the pep8/flake8 skillz to add this to tests so you don't need to keep hand-PEPing? ;P

hashbangstudio replied:
unintentional will alter

ghickman replied:
@hashbangstudio – Thanks!

hashbangstudio committed:
Undo unintentional indent change

hashbangstudio committed:
Update setup.py to inline with master

This should hopefully mean that build will pass with enum dependency met

doismellburning commented on minecraft/block.py:
```diff
@@ -17,26 +14,55 @@ def __init__(self, type=3, data=0):
+    def __eq__(self, rhs):
+        """
+            Equality override
+            Two blocks are equal only if their hashes are equal
```
Except that's only true in this case where `hash` doesn't collide, and not generally true - would you mind making a proper `__eq__` please?

doismellburning commented on minecraft/block.py:
```diff
@@ -1,15 +1,12 @@
     Minecraft PI block description. Can be sent to Minecraft.setBlock/s
-    block.type = the blockID of a block (It's meterial)
-    block.data = A unknown member that does something
+    block.type = the blockID of a block (It's material)
```
If you're fixing the spelling could you fix the apostrophe too please!

doismellburning replied:
(yes "it's" technically works, but it's incredibly misleading...)

doismellburning commented on setup.py:
```diff
@@ -14,17 +21,18 @@
+                   'Programming Language :: Python :: 3.3',
+                   'Programming Language :: Python :: 3.4',
+                   ],
+      )
```
:(

doismellburning commented on tests/test_unit_block.py:
```diff
@@ -0,0 +1,121 @@
+
+class TestBlock(unittest.TestCase):
+
+    def testRepresentation(self):
```
PEP8 / Pythonic method names please

doismellburning commented on tests/test_unit_block.py:
```diff
@@ -0,0 +1,121 @@
+        self.assertEqual(blockType.SAPLING,  6)
+        self.assertEqual(blockType.BEDROCK,  7)
+        self.assertEqual(blockType.WATER_FLOWING,  8)
+        self.assertEqual(blockType.WATER,  blockType.WATER_FLOWING)
```
Could you pull this out into a distinct test please? (as it's somewhat different to the other "is the id the one hardcoded here" tests but easy to not notice!)

hashbangstudio committed:
Fix comparison and typo

Fix __eq__ so block only equal if all attributes equal
Fix apostrophe usage

hashbangstudio committed:
Use PEP8 method names and clarify enum test

Move to PEP8 lowercase_name method names
Move test of water and lava aliases out to separate test

hashbangstudio committed:
Fix missing func def and class naming

Add function (self): that was left missing
blockType changed to BlockType

hashbangstudio committed:
Hopefully correct unintentional indent change

hashbangstudio committed:
Move to PEP8 compliant naming

Change functions and variables to lower_case

doismellburning commented on minecraft/block.py:
```diff
@@ -17,26 +15,55 @@ def __init__(self, type=3, data=0):
+        return self.type
 
 
-class blockType(Enum):
```
I think this changes a bit of the API we don't want to change, but we aren't testing thoroughly enough :s

hashbangstudio commented:
It will need to be changed for PEP8 naming and can't see where it is used in the rest of the API at the moment, unless I am just not seeing it. Can drop it from PR if it is used somewhere.

hashbangstudio committed:
Move to py.test style

Remove unittest style file
Add py.test style file

hashbangstudio committed:
Rename for clarity and PEP8

Variable missed in conversion to PEP8 standards
Change variable name to avoid confusion between withData(data)
  and the Block(type, data) created instances

hashbangstudio committed:
Add eval repr test and use constant string

Add comparison test using eval(repr())
Change to comparing against a constant string

doismellburning commented:
So, sorry about this, but I had to revert a bunch of other changes to fix compatibility breakages they introduced. Are you ok to fix the merge conflicts here, or would you like a hand? Cheers, apologies!

hashbangstudio commented:
Will have a look to see if can fix conflicts over the weekend, if not will let you know

doismellburning commented:
Awesome, thanks, sorry again!

hashbangstudio committed:
Merge remote-tracking branch 'upstream/block-update' into block-update

hashbangstudio committed:
Merge remote-tracking branch 'upstream/master' into block-update

hashbangstudio committed:
Fix conflict between files

both added test_block.py.

Combined in one file

hashbangstudio committed:
Add new block elements

Add BlockType enum back in
Maintain backwards compatibility

doismellburning commented:
<3 thanks for the fixes!

doismellburning commented on mcpi/block.py:
```diff
@@ -1,21 +1,65 @@
+    For example grass, sand, dirt or wool
+    block.data = The variant of the type of block.
+    For example the colour of wool or the orientation of stairs
+    The default type for blocks is dirt
```
Maybe I'm a bit too keen on building obsidian houses, but I'm personally very wary about introducing defaults for what is a somewhat arbitrary choice...

hashbangstudio committed:
Remove default type assignment

hashbangstudio committed:
Also remove the comment line

remove docstring comment on default type

hashbangstudio commented:
Agree is arbitrary, was left in from previous API refactoring. Have removed default assignment.
